### DISCUSSION TITLE
Installed Packs are not visible if Pack Index Cannot be Downloaded Due to Airgapped Environment.

### DISCUSSION THREAD

**Discussion by rpdeo** [StackStorm/st2#6195] [UI]
Hello,

I have a scenario where we have an airgapped deployment and custom packs are uploaded in packs.dev for deployment.
I noticed that the /packs UI route does not show the packs UI if the API cannot obtain index.json from index.stackstorm.org.
Can this behaviour be changed ? I would have expected it to show the already installed packs in the environment, even if the index cannot be loaded.

Please let me know what is the best way to handle this.

**Comment by fdrab** [comment]
This is not a UI issue, the UI is just calling the ST2 packs API. You can host your own packs index https://docs.stackstorm.com/packs.html#hosting-your-own-pack-index

This should solve your issues, I think.

**Reply by rpdeo** [reply]
```
st2-docker-st2api-1  | 2024-05-03 19:40:07,386 ERROR [-] Failed to call controller function "get_all" for operation "st2api.controllers.v1.packs:packs_controller.index.get_all": No result>
st2-docker-st2api-1  | Status: []
st2-docker-st2api-1  | Traceback (most recent call last):
st2-docker-st2api-1  |   File "/opt/stackstorm/st2/lib/python3.8/site-packages/st2common/router.py", line 637, in __call__
st2-docker-st2api-1  |     resp = func(**kw)
st2-docker-st2api-1  |   File "/opt/stackstorm/st2/lib/python3.8/site-packages/st2api/controllers/v1/packs.py", line 354, in get_all
st2-docker-st2api-1  |     index, status = packs_service.fetch_pack_index(proxy_config=proxy_config)
st2-docker-st2api-1  |   File "/opt/stackstorm/st2/lib/python3.8/site-packages/st2common/services/packs.py", line 186, in fetch_pack_index
st2-docker-st2api-1  |     raise ValueError(
st2-docker-st2api-1  | ValueError: No results from the indexes: tried .
st2-docker-st2api-1  | Status: []
st2-docker-st2api-1  | 2024-05-03 19:40:07,387 INFO [-] X-X-X-X-X - 400 66 39.593ms (method='GET',path='/v1/packs/index',remote_addr='X.X.X.X',status=400,runtime=39.593,content_length=66,request_id='X-X-X-X-X')
```

This is the error I am getting in `st2api` logs, the runtime is short (about 40ms) if I unset the `index_url` via `st2.docker.conf` as below. By letting it point to default repo, the runtime increases to 300ms, until the HTTPS/TCP connection attempts time out, and instead of ValueError, it will be HTTPSConnectionError from urllib.

```
[content]
packs_base_paths=/opt/stackstorm/packs.dev
index_url =
```

I am going to try the internal repo approach by hosting a simple nginx webserver for it, will report back on that.

Further, in browser console, I can see xhr entries retrieve the installed pack data, but for some reason, the packs UI dom is not getting built, and my first reaction was, is it because the above API error I see in the `st2api` logs ?

What can I do to debug the web UI part ?

**Reply by chain312** [reply]
If you cannot retrieve the package index and return 400, you can consider adding a proxy in the package configuration so that you can access index.stackstorm.org 's index.json
![image](https://github.com/StackStorm/st2/assets/30886624/91173132-5214-40b5-9980-ad418edf3e52)
Another way is to build a gitlab in the local docker environment, upload the packs application package to gitlab, and configure index_url = as the local address.
![image](https://github.com/StackStorm/st2/assets/30886624/6bf417e1-03b5-40ec-866c-c6aee4a82832)
